### PR TITLE
Do not consider bare TypeVar not overlapping with None for reachability analysis

### LINKED ISSUE #18126: False positive "Statement is unreachable" when checking if typevar is None
**Bug Report**

Mypy doesn't seem to like it when you check if an argument with typevar type is checked to see if it is `None`. It incorrectly reports ` error: Statement is unreachable  [unreachable]`

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.12&flags=strict%2Cwarn-unreachable&gist=ceb8193f19f8b53a55dea70fad802469

```python
import typing as t

T = t.TypeVar("T")

def foo(arg: T) -> None:
    if arg is None:
        return None
    # Do something with arg
    return None


foo(None) ## Allowed
foo(123)
```

Run mypy with `--warn-unreachable` flag

**Expected Behavior**

No errors

**Actual Behavior**

```
$ mypy .vscode/scratch.py
.vscode/scratch.py: note: In function "foo":
.vscode/scratch.py:9: error: Statement is unreachable  [unreachable]
Found 1 error in 1 file (checked 1 source file)
```

The error seems to only happen when checking if arg is `None`, but not with other types. For example, I tired changing `if arg is None:` line to the following and these were the results.

1. `if arg is True:` -> no errors
2. `if isinstance(arg, int):` -> no errors
3. `if isinstance(arg, type(None)):` -> unreachable error
4. `if arg == None:` -> No error

Also, changing the type to `T | None` removes the error as well.

**Your Environment**

- Mypy version used: mypy 1.13.0 (compiled: yes)
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files): `warn_unreachable = True`
- Python version used: 3.12

### PR DESCRIPTION
**Pull request by sterliakov**
Fixes #18126.

Simply allowing such intersection was insufficient: existing binder logic widened the type to `T | None` after the `is None` check.

This PR extends the binder logic to prevent constructing a union type when all conditional branches are reachable and contain no assignments: checking `if isinstance(something, Something)` does not change the type of `something` after the end of the `if` block.

### CODE REVIEW

**Comment by sterliakov** [comment]
CI failed with the Windows problem I faced multiple times before, the rest is green.

Primer diff is amazing this time! Every diff line seems to be an improvement.

* `jinja2`: unrelated meet-vs-join problem. Minimal repro ([playground](https://mypy-play.net/?mypy=1.13.0&python=3.12&flags=strict&gist=f2df47d37c231b7d35e26744191d9e11)):
  ```python
  class A: pass
  class B: pass

  foos: tuple[A, ...] | tuple[B, ...]

  for _, foo in enumerate(foos):
      reveal_type(foo)  # N: Revealed type is "builtins.object"
  ```
* `prefect`: the error is correct. I don't know why it wasn't emitted previously, though: there's assignment of different types to an unannotated variable in if/else branches.
* `sphinx/search/__init__.py`: bug in `docutils-stubs` (`children` is declared in [`Node`](https://github.com/docutils/docutils/blob/ceb8eb76a21da5553017bedb9eca8bb18ebde57f/docutils/docutils/nodes.py#L81), but not in [stubs](https://github.com/tk0miya/docutils-stubs/blob/b14071d065f8ee35399856fcaff6cd608294a342/docutils-stubs/nodes.pyi#L17))
* `sphinx/ext/graphviz.py`: correct! This attribute isn't declared on any of the classes. Still don't know how it relates to this change, though...
* `schema_salad`: correct! `self.name` is `Any | None` and is passed to a function expecting `str`.
* `kopf`: another unrelated bug:( Minimal repro ([playground](https://mypy-play.net/?mypy=master&python=3.12&flags=strict&gist=10d9f159ae3dcb3943f39e8720c13ddd) with sync/async comparison):
  ```python
  from typing import Any, Callable, TypeVar

  T = TypeVar("T")

  async def call(fn: Callable[[], T]) -> T:
      return fn()

  make_x: Any

  async def fn() -> None:
      x: Any | None = None
      x = await call(make_x)
      reveal_type(x)  # N: Revealed type is "Union[Any, None]"
  ```
* `pytest`: correct. There's really no error.
* `paroxython`: correct, there's a narrowing not supported by `mypy`
* `pandera`: I don't understand what's type-ignored there...
* `cki-lib`: correct! It boils down to the following:
  ```python
  x: Any
  if x is None:
      ...
  do_something_with_not_none(x)
  ```
* `pygithub`: correct! That is really a type error.
* `discord`: all correct! Especially nice to get rid of `Item "object" of the upper bound "Cog | None"` - that was strange.
* `steam`: both correct! Second one suffered from needless type widening after `isinstance`
* `ibis`: all correct!

File: mypy/binder.py [hunk 1]
```diff
@@ -236,8 +257,8 @@ def update_from_options(self, frames: list[Frame]) -> bool:
                         )
                         if simplified == self.declarations[key]:
                             type = simplified
-            if current_value is None or not is_same_type(type, current_value):
-                self._put(key, type)
+            if current_value is None or not is_same_type(type, current_value[0]):
+                self._put(key, type, True)
```
Comment (JukkaL) [inline mypy/binder.py:261]: Nit: use `from_assignment=True` to make this a little clearer?

File: mypy/checker.py [hunk 2]
```diff
@@ -4725,11 +4725,11 @@ def visit_if_stmt(self, s: IfStmt) -> None:
 
                 # XXX Issue a warning if condition is always False?
                 with self.binder.frame_context(can_skip=True, fall_through=2):
-                    self.push_type_map(if_map)
+                    self.push_type_map(if_map, False)
```
Comment (JukkaL) [inline mypy/checker.py:4728]: Also consider using `from_assignment=False` here and elsewhere in this file (when just literal `True`/`False` is passed).

**Review by JukkaL — APPROVED** [review]
Nice, based on mypy_primer output this fixes multiple bugs, and makes the inferred types more consistent. Just left some minor comments about style (optional), looks good otherwise.

**Review by JukkaL — APPROVED** [review]
Thanks for the updates, LGTM!

**Comment by PetterS** [comment]
Very nice!

**Pull request merged**